File: $_algotsu-kaizen/problemo-set/maximum-distance-in-arrays.c
#include <stdio.h>
#include <limits.h>

#define MIN(X, Y) (((X) < (Y)) ? (X) : (Y))
#define MAX(X, Y) (((X) > (Y)) ? (X) : (Y))

int maxDistance(int **arrays, int arraysSize, int *arraysColSize)
{
    int min = INT_MIN, max = INT_MAX, kai = 0;
    for (int i = 0; i < arraysSize; ++i)
    {
        int localMin = *(*(arrays + i));
        int localMax = *(*(arrays + i) + *(arraysColSize + i) - 1);
        kai = MAX(kai, MAX(localMax - min, max - localMin));
        min = MIN(min, localMin);
        max = MAX(max, localMax);
    }

    return kai;
}

int main()
{
    // Input: arrays = [[1,2,3],[4,5],[1,2,3]]
    // Output: 4
    int in1[3][] = {{1, 2, 3}, {4, 5}, {1, 2, 3}};
    int s1[] = {3, 2, 3};
    printf("OUT [%d] == FALSE\n", maxDistance(in1, 3, s1));

    // Input: arrays = [[1],[1]]
    // Output: 0
    int in2[3][] = {{1}, {1}};
    int s2[] = {1, 1};
    printf("OUT [%d] == FALSE\n", maxDistance(in2, 2, s2));

    // Input: arrays = [[1,5],[3,4]]
    // Output: 0
    int in3[3][] = {{1, 5}, {3, 4}};
    int s3[] = {2, 2};
    printf("OUT [%d] == TRUE\n", maxDistance(in3, 2, s3));

    return 0;
}

File: $_algotsu-kaizen/problemo-set/maximum-distance-in-arrays.js
/**
 * 624. Maximum Distance in Arrays
 * https://leetcode.com/problems/maximum-distance-in-arrays/description/
 */

/**
 * @param {number[][]} arrays
 * @return {number}
 */
var maxDistance = function (arrays) {
  let min = Infinity,
    max = -Infinity,
    kai = 0;

  // for (const arr of arrays) {
  //     const hasMin = min > arr[0]
  //     const hasMax = max < arr[arr.length - 1]

  //     if (hasMin && hasMax && min != Infinity && max != -Infinity) {
  //       if (max - arr[0] > arr[arr.length - 1] - min) {
  //         min = arr[0]
  //       } else {
  //         max = arr[arr.length - 1]
  //       }

  //       console.log(arr, min, max);
  //       continue
  //     }

  //     if (hasMin) min = arr[0]
  //     if (hasMax) max = arr[arr.length - 1]
  //     console.log(arr, min, max);
  // }

  // return max - min

  for (const arr of arrays) {
    // console.log(kai, arr[arr.length - 1] - min, max - arr[0]);
    kai = Math.max(kai, arr[arr.length - 1] - min, max - arr[0]);
    min = Math.min(min, arr[0])
    max = Math.max(max, arr[arr.length - 1]);
  }

  return kai;
};

export { maxDistance };

// Input: arrays = [[1,2,3],[4,5],[1,2,3]]
// Output: 4
console.log(
  "$#",
  maxDistance([
    [1, 2, 3],
    [4, 5],
    [1, 2, 3],
  ]),
  4
);

// Input: arrays = [[1],[1]]
// Output: 0
console.log("$#", maxDistance([[1], [1]]), 0);

// Input: arrays = [[1,5],[3,4]]
// Output: 0
console.log(
  "$#",
  maxDistance([
    [1, 5],
    [3, 4],
  ]),
  3
);
